src/rgw/rgw_sal.cc

```diff
@@ -93,213 +83,86 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
+    if (newDriver)
+      store = newDriver(dpp, cct, false, use_gc_thread, use_lc_thread, quota_threads, run_sync_thread, run_reshard_thread, use_cache, use_gc);
+    if (dlclose(dl) < 0)
+      ldpp_dout(dpp, 0) << "WARNING: dlclose() failed" << dendl;
   }
-#endif
 
   if (cfg.filter_name.compare("base") == 0) {
-    rgw::sal::Driver* next = driver;
-    driver = newBaseFilter(next);
+    next = store;
+    store = newBaseFilter(next);
+  }
 
-    if (driver->initialize(cct, dpp) < 0) {
-      delete driver;
-      delete next;
-      return nullptr;
-    }
+  if (store->initialize(cct, dpp) < 0) {
+    delete store;
+    delete next;
+    return nullptr;
   }
 
-  return driver;
+  return store;
 }
 
-rgw::sal::Driver* DriverManager::init_raw_storage_provider(const DoutPrefixProvider* dpp, CephContext* cct, const Config& cfg)
+rgw::sal::Driver* DriverManager::init_raw_storage_provider(const DoutPrefixProvider* dpp,
+                                                           CephContext* cct,
+                                                           const Config& cfg)
 {
-  rgw::sal::Driver* driver = nullptr;
-  if (cfg.store_name.compare("rados") == 0) {
-    driver = newRadosStore();
-    RGWRados* rados = static_cast<rgw::sal::RadosStore* >(driver)->getRados();
-
-    rados->set_context(cct);
-
-    int ret = rados->init_svc(true, dpp);
-    if (ret < 0) {
-      ldout(cct, 0) << "ERROR: failed to init services (ret=" << cpp_strerror(-ret) << ")" << dendl;
-      delete driver;
-      return nullptr;
-    }
-
-    if (rados->init_rados() < 0) {
-      delete driver;
-      return nullptr;
-    }
-    if (driver->initialize(cct, dpp) < 0) {
-      delete driver;
-      return nullptr;
-    }
-  } else if (cfg.store_name.compare("dbstore") == 0) {
-#ifdef WITH_RADOSGW_DBSTORE
-    driver = newDBStore(cct);
-
-    if ((*(rgw::sal::DBStore*)driver).initialize(cct, dpp) < 0) {
-      delete driver;
-      return nullptr;
-    }
-#else
-    driver = nullptr;
-#endif
-  } else if (cfg.store_name.compare("motr") == 0) {
-#ifdef WITH_RADOSGW_MOTR
-    driver = newMotrStore(cct);
-#else
-    driver = nullptr;
-#endif
-  } else if (cfg.store_name.compare("daos") == 0) {
-#ifdef WITH_RADOSGW_DAOS
-    driver = newDaosStore(cct);
-
-    if (driver->initialize(cct, dpp) < 0) {
-      delete driver;
-      return nullptr;
-    }
-#else
-    driver = nullptr;
-#endif
+  auto dlname = fmt::format("/usr/lib64/ceph/librgw_sal_{}.so", cfg.store_name);
+  rgw::sal::Driver *store = nullptr;
+    rgw::sal::Driver *next = nullptr;
+  void *dl = nullptr;
+  rgw::sal::Driver *(*newDriver)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool) = nullptr;
+  dl = dlopen(dlname.c_str(), RTLD_NOW | RTLD_LOCAL | RTLD_DEEPBIND);
+  if (dl) {
+    newDriver = (rgw::sal::Driver* (*)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool))dlsym(dl, "new_Driver");
```

Comment:
the function pointer deserves a typedef in rgw_sal.h
```c++
using rgw_sal_create_fn = rgw::sal::Driver* (*)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool);
```
and factory functions should return `std::unique_ptr` instead of raw pointers

Reply:
function ptr in rgw_sal.h will be in next push.

making DriverManager::get_storage()  and DriverManager::get_raw_storage() return std::unique_ptr is more invasive and will be in a follow-on push. (Maybe Dan can did this faster than I can?)

Reply:
no problem, we can follow up on that later